server/src/main/java/org/folio/shared/index/api/OaiService.java

```diff
@@ -0,0 +1,374 @@
+          break;
+        }
+      }
+      if (cmp == 0) {
+        return field.getJsonArray("subfields");
+      } else if (cmp < 0) {
+        break;
+      }
+    }
+    JsonObject field = new JsonObject();
+    fields.add(i, new JsonObject().put(tag, field));
+    field.put("ind1", ind1);
+    field.put("ind2", ind2);
+    JsonArray subfields = new JsonArray();
+    field.put("subfields", subfields);
+    return subfields;
+  }
+
+  static void parseHoldingsRecords(JsonArray field, JsonObject inventoryPayload) {
+    if (inventoryPayload == null) {
+      return;
+    }
+
+    JsonArray holdingsRecords = inventoryPayload.getJsonArray("holdingsRecords");
+    if (holdingsRecords == null) {
+      return;
+    }
+    for (int i = 0; i < holdingsRecords.size(); i++) {
+      JsonObject holdingsRecord = holdingsRecords.getJsonObject(i);
+      String permanentLocation = holdingsRecord.getString("permanentLocationDeref");
+      if (permanentLocation != null) {
+        field.add(new JsonObject().put("b", permanentLocation));
+      }
+    }
+  }
+
+  static String getMetadata(RowSet<Row> rowSet, UUID clusterId, List<String> matchValues) {
+    RowIterator<Row> iterator = rowSet.iterator();
+    JsonObject marc = null;
+    JsonArray holdings = new JsonArray();
+    JsonArray identifiersField = new JsonArray();
+    identifiersField.add(new JsonObject().put("i", encodeOaiIdentifier(clusterId)));
+    for (String matchValue : matchValues) {
+      identifiersField.add(new JsonObject().put("m", matchValue));
+    }
+    while (iterator.hasNext()) {
+      Row row = iterator.next();
+      if (marc == null) {
+        marc = row.getJsonObject("marc_payload");
+      }
+      JsonObject inventoryPayload = row.getJsonObject("inventory_payload");
+      parseHoldingsRecords(holdings, inventoryPayload);
+      identifiersField.add(new JsonObject()
+          .put("l", row.getString("local_id"))
+          .put("s", row.getUUID("source_id").toString())
+      );
+    }
+    if (marc == null) {
+      return null; // a deleted record
+    }
+    clearMarcField(marc, "999");
+    updateMarcSubFields(marc, "999", "1", "0").addAll(identifiersField);
+    JsonArray f852 = updateMarcSubFields(marc, "852", "0", " ");
+    f852.clear();
+    f852.addAll(holdings);
+    String xmlMetadata = XmlJsonUtil.convertJsonToMarcXml(marc);
+    return "    <metadata>\n" + xmlMetadata + "\n    </metadata>\n";
+  }
+
+  static Future<String> getXmlRecordMetadata(Storage storage, SqlConnection conn, UUID clusterId,
+      List<String> matchValues) {
+    String q = "SELECT * FROM " + storage.getBibRecordTable()
+        + " LEFT JOIN " + storage.getClusterRecordTable() + " ON record_id = id "
+        + " WHERE cluster_id = $1";
+    return conn.preparedQuery(q)
+        .execute(Tuple.of(clusterId))
+        .map(rowSet -> getMetadata(rowSet, clusterId, matchValues));
+  }
+
+  static Future<List<String>> getClusterValues(Storage storage, SqlConnection conn,
```

Comment:
It worked pretty bad with the non-OAI stuff. You see that Storage.java pretty much has everything and the SharedIndexService is small. SO I'd rather keep the things combined but split according to functionality.. Eg MachKeyService, ClusterService, RecordService and now OaiService.

Reply:
Can we split by two axis: by functionality and by layer so that the DB acces code (xStorage), XML/JSON/Other format mapping and HTTP req/res handling is in seperate units?